### ISSUE TITLE
Web - Chat-  A different group chat appears when refreshing the page

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#30868]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.94.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @johnmlee101 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1698904742009159

## Action Performed:
1. Create a group conversation
2. Type something 
3. Clicking on a profile 
4. Refresh the page

## Expected Result:
The same group chat appears when refreshing the page

## Actual Result:
A different group chat appears when refreshing the page

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/93399543/8439bb82-91bd-4cd8-aecb-da7059063b20

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
A different group chat is opened when we refresh while we have opened profile detail of group chat member
### What is the root cause of that problem?
When we open profile detail of a member of a group chat the url is changed to '.../a/memberIDorParticipantID' so the report id is lost from the url, hence when we refresh the last group chat the user was on is not displayed.
### What changes do you think we should make in order to solve the problem?

We need to save last report id the user accessed in onyx everytime a user navigates to a report id then we will navigate it to that report when the profile detail page is open in RHN.
### What alternative solutions did you explore? (Optional)

**Comment by abzokhattab** [comment]
# Proposal
## Please re-state the problem that we are trying to solve in this issue.
A different report  is opened when we refresh while we have opened profile detail of group chat member

## What is the root cause of that problem?

the profile page route doesnt contain any info about the reportId therefore the opened report in the background is not the same 

## What changes do you think we should make in order to solve the problem?
we need to modify the url of the profile page to contain the url of the report: 
for example: 
this way the screenwrapper will get the reportid and then displays it 
`/r/reportId/participants/:id/profile/`

or we can add `reportId` as a param and modify the screener to look for this param and open it in the background 

What alternative solutions did you explore? (Optional)

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Web - Chat- A different group chat appears when refreshing the page
### What is the root cause of that problem?
- When launching the app with the deeplink that does not contain reportID, we use the ```getLastAccessedReportID``` to get the report to display.
https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/Navigation/AppNavigator/ReportScreenIDSetter.js#L63-L72
- In the above, the ```findLastAccessedReport``` is not working properly. 
https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/ReportUtils.js#L575-L581
- ```findLastAccessedReport``` uses ```sortReportsByLastRead```:
https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/ReportUtils.js#L263-L269 that will sort the reports by "lastReadTime". This is incorrect because the "lastReadTime" refers to the latest time the user read a new message, not the latest time the user visits a report. For example, if we read a new message A on 3/10/2023, and re-visit its report on 4/10/2023, the "lastReadTime" is still '4/10/2023'.
### What changes do you think we should make in order to solve the problem?
- Once we call the API 'OpenReport', we need to add the ```lastVisitTime: DateUtils.getDBTime()``` to the ```ONYXKEYS.COLLECTION.REPORT_METADATA```:
https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/actions/Report.js#L476-L484
- Then, in the ```sortReportsByLastRead```, using ```lastVisitTime``` to sort the reports. We can do it by updating:
https://github.com/Expensify/App/blob/7bd51eba84f6708409dfd6e6886bf968eff3358f/src/libs/ReportUtils.js#L263-L269
to:
```
function sortReportsByLastRead(reports, reportMetadata) {
    return _.chain(reports)
        .map((report, reportID) => {
            if (!reportMetadata[reportID]) {
                return report;
            }
            return {...report, ...reportMetadata[reportID]};
        })
        .toArray()
        .filter((report) => report && report.reportID && report.lastReadTime)
        .sortBy('lastReadTime')
        .value();
}
```


### What alternative solutions did you explore? (Optional)
- NA

**Comment by shubham1206agra** [comment]
@FitseTLT Your proposal might not work with deleting reports. Do you have any workaround for this?

**Comment by FitseTLT** [comment]
> @FitseTLT Your proposal might not work with deleting reports. Do you have any workaround for this?

Yes instead of saving a single reportID everytime a user opens a report we can save a stack of reportIDs and when opening the reportIDSetter will populate the recent available reportID to the `ReportScreen` 👍  .

**Comment by FitseTLT** [comment]
@shubham1206agra Updated [proposal](https://github.com/Expensify/App/issues/30868#issuecomment-1792965156)

**Comment by shubham1206agra** [comment]
Since @FitseTLT's solution will not be as strong as I thought. And @DylanDylann's [proposal](https://github.com/Expensify/App/issues/30868#issuecomment-1793304559) will cover more wide cases. I think we should go for @DylanDylann.

🎀 👀 🎀 C+ reviewed

cc @MonilBhavsar

**Comment by FitseTLT** [comment]
@MonilBhavsar  @shubham1206agra Please take a look at this [solution](https://github.com/Expensify/App/issues/30778#issuecomment-1797934564) to this [issue](https://github.com/Expensify/App/issues/30778) they are using an approach that is similar to what I proposed and if they are going to implement it, the updating of `lastVisitTime` approach you have chosen here will be redundant/ overlapping solution to that approach they are using to save a stack of recently opened reports in Onyx. 👍

**Comment by johnmlee101** [comment]
Amending the issue body, @jliexpensify was the reporter of this issue

**Comment by DylanDylann** [comment]
@FitseTLT My proposal fix the root cause that the ```sortReportsByLastRead``` function is working incorrectly.
https://github.com/Expensify/App/blob/dee9b3110a7ebe0aa250c3f966d6a69925d916f4/src/libs/ReportUtils.js#L263-L269

- And the result of this function will be used in many places in our app, and also can use to fix the issue "Drop user back to the last chat they were on when they leave a room" that you are mentioned

**Comment by FitseTLT** [comment]
@DylanDylann I know that your solution solves both issues that's what I commented but @shubham1206agra the advantage of my proposal and the one [proposed](https://github.com/Expensify/App/issues/30778#issuecomment-1797934564) for the other issue is that it is efficient memory-wise to only save a few last accessed report Ids (maybe 5) instead of a `lastVisitTime` field for every single reports a user has. Please try to reconsider it and also communicate with the assignees of that issue 👍 Thanks.
cc @jliexpensify @MonilBhavsar

**Comment by DylanDylann** [comment]
@FitseTLT Yeah, thanks @FitseTLT. I`ve commented on the issue that you mentioned to discuss the expected solution.

**Comment by shubham1206agra** [comment]
So, I have discussed this with @MonilBhavsar. Given https://github.com/Expensify/App/issues/30868#issuecomment-1802566373, we are going to put the issue on HOLD till that issue is resolved, and then we are simply going to use this.

So, I am going to take back my previous review.

**Comment by MonilBhavsar** [comment]
Bumping to weekly as it is on HOLD

**Comment by MonilBhavsar** [comment]
Taking this off hold as we discussed here https://expensify.slack.com/archives/C02NK2DQWUX/p1700121699237619?thread_ts=1698756003.834449&cid=C02NK2DQWUX

@DylanDylann Could you please share test branch so we can test the proposal. We're thinking to move forward with that solution to keep behavior same as of today and not over engineer it. Thanks!

**Comment by DylanDylann** [comment]
@MonilBhavs Here is the draft PR https://github.com/Expensify/App/pull/31424

**Comment by shubham1206agra** [comment]
Sorry for the delay. I will update here as soon as possible.

**Comment by techievivek** [comment]
@DylanDylann Hi, how is the PR going? When can we expect it to be open for review? We have put this PR https://github.com/Expensify/App/pull/31097 on HOLD for the changes being added here. Let us know if you need any help in moving it forward.

**Comment by DylanDylann** [comment]
@techievivek I`ve created the draft PR and request @shubham1206agra review https://github.com/Expensify/App/issues/30868#issuecomment-1814077551

**Comment by shubham1206agra** [comment]
I have reviewed it. Nothing seems broken to me right now.

@MonilBhavsar Please assign the issue to @DylanDylann.

And @DylanDylann can you merge main there? And we can create adhoc builds to test performance issue on physical devices too.

**Comment by abekkala** [comment]
I'm heading out for the Thanksgiving holiday and will be back bright and early Monday morning. 
If something pops up that needs urgent attention please reapply label or ask in slack. 

@DylanDylann can you note when you expect a PR to be ready?

**Comment by DylanDylann** [comment]
@abekkala I am wokring on it and I will open the PR today

**Comment by DylanDylann** [comment]
@shubham1206agra PR https://github.com/Expensify/App/pull/31424 is ready to reivew

**Comment by abekkala** [comment]
note: PR hasn't merged yet.

**Comment by shubham1206agra** [comment]
@abekkala There are some discussion going on about performance regressions in the PR

**Comment by abekkala** [comment]
@shubham1206agra yes, I was just making an internal note for myself. thanks

**Comment by abekkala** [comment]
@lschurr 
### STATUS 
I'm ooo through Jan 03. 

This is already at the PR stage, but encountering some issues. 

If this actually reaches payment date before I return:

- Issue reported by: @jliexpensify - no payment 😉 
- Fix:  @DylanDylann
- PR Review: @shubham1206agra

**Comment by lschurr** [comment]
It looks like [this PR](https://github.com/Expensify/App/pull/31424) still isn't merged to prod. I believe we're waiting on that before the payment date will be set.

**Comment by lschurr** [comment]
Leaving this with you @abekkala since I think you'll be back for the payout date!

**Comment by shubham1206agra** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@shubham1206agra] The PR that introduced the bug has been identified. Link to the PR: None as this is new functionality
* [x]  [@shubham1206agra] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: NA
* [x]  [@shubham1206agra] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: Not required
* [x]  [@shubham1206agra] Determine if we should create a regression test for this bug. Yes
* [x]  [@shubham1206agra] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

1. Navigate to a group conversation
2. Clicking on a profile
3. Refresh the page. Check the background report opened.

* [ ]  [@abekkala / @lschurr] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by abekkala** [comment]
@shubham1206agra I think we may need to get a bit more crisp on your regression test

> 1. Navigate to a group conversation
> 2. Clicking on a profile
> 3. Refresh the page. Check the background report opened.


Based on the OP

### Action Performed:

1. Create a group conversation
2. Type something
3. Clicking on a profile
4. Refresh the page

### Expected Result:
The same group chat appears when refreshing the page

in your regression would we need to type and/or send a message prior to clicking on a profile. After refreshing the page the last step should be "verifying" the correct result. Meaning, should it be "**Verify** the group chat still appears" rather than the issue of a different group chat appearing after refreshing the page

**Comment by shubham1206agra** [comment]
@abekkala Updated the test.

**Comment by abekkala** [comment]
1. Navigate to a group conversation
2. Click on a profile
3. Refresh the page
4. Verify the background report opened is same as group conversation opened in step 1.

**Comment by abekkala** [comment]
@shubham1206agra payment sent and contract ended - thank you! 🎉

**Comment by abekkala** [comment]
@DylanDylann payment sent and contract ended - thank you! 🎉
